lib/phoenix_live_view/engine.ex

```diff
@@ -871,23 +918,37 @@ defmodule Phoenix.LiveView.Engine do
     dynamic |> Map.merge(static) |> Map.put(:__changed__, changed)
   end
 
-  defp component_changed(:all, _assigns, _changed, _vars_changed), do: true
+  defp component_changed(:all, _assigns, _changed, _vars_changed_vars, _vars_changed), do: true
 
-  defp component_changed([path], assigns, changed, vars_changed) do
+  defp component_changed([path], assigns, changed, vars_changed_vars, vars_changed) do
     case path do
-      {:changed, [key]} -> changed_assign(changed, key)
-      {:changed, [key | tail]} -> nested_changed_assign(tail, key, assigns, changed)
-      {:vars_changed, [key]} -> changed_assign(vars_changed, key)
-      {:vars_changed, [key | tail]} -> nested_changed_assign(tail, key, assigns, vars_changed)
+      {:changed, [key]} ->
+        changed_assign(changed, key)
+
+      {:changed, [key | tail]} ->
+        nested_changed_assign(tail, key, assigns, changed)
+
+      {:vars_changed, [key]} ->
+        changed_assign(vars_changed, key)
+
+      {:vars_changed, [key | tail]} ->
+        nested_changed_assign(tail, key, vars_changed_vars, vars_changed)
```

Comment:
This is the important part. Previously, the nested case looked into the regular assigns, which of course doesn't work properly and leads to invalid cases when there actually is an assign with the same name.